tests/unit/igraph_coloring.c

```diff
@@ -0,0 +1,56 @@
+#include <igraph.h>
+
+#include "test_utilities.h"
+
+igraph_error_t verify_coloring(igraph_t *graph, igraph_vector_int_t *colors){
+    /* Verify that the colouring is valid, i.e. no two adjacent vertices have the same colour. */
+    igraph_integer_t i;
+    /* Store the edge count to avoid the overhead from igraph_ecount in the for loop. */
+    igraph_integer_t no_of_edges = igraph_ecount(graph);
+    for (i = 0; i < no_of_edges; ++i) {
+        if ( VECTOR(*colors)[ IGRAPH_FROM(graph, i) ] == VECTOR(*colors)[ IGRAPH_TO(graph, i) ]  ) {
+            printf("Inconsistent coloring! Vertices %" IGRAPH_PRId " and %" IGRAPH_PRId " are adjacent but have the same color.\n",
+                   IGRAPH_FROM(graph, i), IGRAPH_TO(graph, i));
+            return IGRAPH_FAILURE;
+        }
+    }
+    igraph_integer_t vertex_count = igraph_vcount(graph);
+    /*verify all assigned colors are valid*/
+    for(i = 0 ; i < vertex_count ; i++){
+        igraph_integer_t color = VECTOR(*colors)[i];
+        if(color<0 || color>=vertex_count){
+            printf("The vertex %" IGRAPH_PRId " has invalid color %" IGRAPH_PRId "\n", i, color);
+            return IGRAPH_FAILURE;
+        }
+    }
+    return IGRAPH_SUCCESS;
+}
+
+int main(void) {
```

Comment:
So far there's a test done on a very large random graph. That's great. We need a few more test cases for confidence though:

 - Add a few small test graphs, something small enough that it should be possible to verify the result manually. Print out the colouring, and create an `.out` file to verify it.  You'll find more information in https://github.com/igraph/igraph/wiki/How-to-write-unit-tests%3F
 - Test the function on the zero-vertex and one-vertex graphs.
 - Test the function on non-simple graphs (containing multi-edges and self-loops).

Reply:
Writing a .out file for colorings does not seem like a feasible idea. Since coloring is an NP Complete problem there is no correct solutions and different algorithms will come to different solutions. A coloring is a set partition is labelling is assigning a number to said paritions so the exact same coloring can have different labellings as well, so I just don't know how i would go about making a .out file.

Reply:
Here the purpose is not to verify that the colouring is valid. We have a function for that. The aim is to detect if the algorithm gets broken in the future. Consider e.g. how the first version in this PR considered the degree in the entire graph rather than just thenon-coloured part of the graph.

 - Let the program print the colour assigned to each vertex
 - Record the output into an `.out` file

This will help detect if the code is inadvertently broken in the future.

Reply:
Ok. I will created a .out file